Chapter: Related Work / Section: Language Servers (#16)
ysndr/MSc-Thesis: chapter/related-work.md

## Language Servers

The LSP project was announced [@lsp-announced] in 2016 to establish a common protocol over which language tooling could communicate with editors.

> This way it is able to fully concentrate on source analysis instead of integration with specific editors by building custom GUI elements and being restricted to editors extension interface.

Reviewer: Which way? what is "it"? I would suggest to be more explicit: "The LSP helps the language intelligence tooling to fully concentrate on source code analysis, independently of the user interface and the other peculiarities of each editor."

At the time of writing the LSP is available in version `3.16` [@Spec].
Microsoft's official website lists 172 implementations of the LSP[@implementations] for an equally impressive number of languages.

> Assessing any majority of these is outside the scope of this work.

Reviewer: This sentence sounds slightly odd. Is there a missing word somewhere? What does "any majority" means?

Yet, a few implementations stand out due to their sophisticate architecture, features, popularity or closeness to the presented work.

### Considerable dimensions

To be able to compare and describe each project objectively and comprehensively, the focus will be on the following dimensions.

Target Language
  ~ The complexity of implementing language servers is influenced severely by the targeted language.

>     Feature rich languages naturally require more sophisticated solutions than simpler ones.

Reviewer: ```suggestion
    Feature rich languages naturally require more sophisticated.
```
Just a suggestion. The second part is a tad redundant.

>     Conversely, self-hosted and interpreted languages can often leverage existing tooling.

Reviewer: I'm not sure to understand the "conversely": what previous statement does it oppose? Otherwise, why does the language being self-hosted or interpreted make a difference?

Author: Lets be more general here then instead

Features
  ~ The LSP defines an extensive array of capabilities.
    The implementation of these protocol features is optional and servers and clients are able to communicate a set of *mutually supported* capabilities.

>   ~ The Langserver.org^[https://langserver.org] project identified six basic capabilities that are most widely supported:

Author: ```suggestion
  ~ The Langserver [@langserver.org] project identified six basic capabilities that are most widely supported:
```

>     1. `Code completion`
>     2. `Hover`,
>     3. `Jump to def`,
>     4. `Workspace symbols`,
>     5. `Find references`
>     6. Diagnostics

Author: ```suggestion
    1. Code completion,
    2. Hover information,
    3. Jump to definition,
    4. Find references,
    5. Workspace symbols,
    6. Diagnostics
```

>     6. Diagnostics

Reviewer: ```suggestion
    6. `Diagnostics`
```
Not sure, but the others got their backticks.

Author: ```suggestion
    1. Code completion,
    2. Hover information,
    3. Jump to definition,
    4. Find references,
    5. Workspace symbols,
    6. Diagnostics
```

  ~ Yet, not all of these are applicable in every case and some LSP implementations reach for a much more complete coverage of the protocol.

File Processing

>   ~ Most language servers use very different methods of handling the source code they analyze.

Reviewer: Nitpick, but example of making the two lines slightly lighter:

"Different language servers may handle source code analysis in very different ways.
 The complexity of the language influences the choice of an approach."

>   ~ Most language servers use very different methods of handling the source code they analyze.
>     The means are mainly influenced by the complexity of the language.

Author: ```suggestion
  ~ Most language servers handling source code analysis in different ways.
    The complexity of the language can be a main influence for the choice of the approach.
```

Author: In response to:

> Nitpick, but example of making the two lines slightly lighter:

> "Different language servers may handle source code analysis in very different ways. The complexity of the language influences the choice of an approach."

    Distinctions appear in the way servers process *file indexes and changes* and how they respond to *requests*.
  ~ The LSP supports sending updates in form of diffs of atomic changes and complete transmission of changed files.

>     The former requires methods of incremental parsing and analysis which are difficult by itself, but are able to process files much more quickly.

Reviewer: ```suggestion
    The former requires incremental parsing and analysis, which are challenging to implement but make processing files much faster upon changes.
```

>     An incremental approach makes use of an internal representation of the source code that allows to quickly derive analytic results from and can be updated efficiently.

Reviewer: I would say the quickly derive analytic results is secondary (and common to all implementation, incremental or not), so I would insist on the "being updated efficiently upon small changes in the source file" (if only by putting it in first position in the rest of the sentence).

Author: ```suggestion
    An incremental approach makes use of an internal representation of the source code that allows efficient updates upon small changes to the source file.

```

>     Additionally, to facilitate the parsing, it must be able to provide a parser with the right context to correctly parse a changed fragment of code.

Reviewer: what does "it" refers to? Also there's a lot of "parse" in this sentence, it reads tediously (repetition is fine in English, but maybe this one could be slightly compressed or reworded?)

Author: ```suggestion
    Additionally, to facilitate the parsing, an incremental approach must be able to provide a parser with the right context to correctly parse a changed fragment of code.
```

>     On the contrary, most language servers process file changes by re-indexing the entire file, updating their internal model much more broadly.

Reviewer: ```suggestion
   In practice, most language servers process file changes by re-indexing the entire file, updating their internal model much more broadly.
```

"updating their internal model much more broadly": what does it mean exactly?

Author: Technically, "replace", ie. all of the model is updated. Incremental ones update the model only locally and might invalidate some caches that get recomputed lazily

Reviewer: Then, I think using a sentence with "replace" is better. Or I would even say that the second part of the sentence is not that informative, as "re-indexing the entire file" does seem to suggest that all the work is done from zero again.

Author: How about this: stressing that the model gets discarded?

Reviewer: Yep, looks good to me

>     This is a more approachable method.
>     Yet, it is less performant since entire files need to be processed, which becomes more noticeable as file sizes and edit frequency increase.

Author: ```suggestion
    This is a more approachable method, as it poses less requirements to the architects of the language server.
    Yet, it is far less performant.
    Unlike incremental processing (which updates only the affected portion of its internal structure), the smallest changes, including adding or removing lines effect the _reprocessing of the entire file_.
    While sufficient for small languages and codebases, non-incremental processing quickly becomes a performance bottleneck.
    
```

>     Yet, it is less performant since entire files need to be processed, which becomes more noticeable as file sizes and edit frequency increase.

Reviewer: I think you can emphasize much more the aspect that incrementality is about *small changes imply small reprocessing* in the paragraph above. For example, entire files also need to be processed by incremental LSP, but the point is that they don't need to be entirely REprocessed upon small changes.

  ~ For code analysis LSP implementers have to decide between *lazy* or *greedy* approaches for processing files and answering requests.
    Dominantly greedy implementations resolve most available information during the indexing of the file.

>     The server can then facilitate this model to answer requests using mere lookups.

Reviewer: facilitate this model? what does it mean?
